ext/opcache/jit/zend_jit_ir.c

```diff
@@ -8713,15 +8714,28 @@ static int zend_jit_push_call_frame(zend_jit_ctx *jit, const zend_op *opline, co
 		// JIT: Z_PTR(call->This) = object_or_called_scope;
 		ir_STORE(jit_CALL(rx, This.value.ptr), object_or_called_scope);
 
-		// JIT: if (closure->func.op_array.run_time_cache__ptr)
-		if_cond = ir_IF(ir_LOAD_A(ir_ADD_OFFSET(func_ref, offsetof(zend_closure, func.op_array.run_time_cache__ptr))));
-		ir_IF_FALSE(if_cond);
+		if (!func) {
+			// JIT: if (closure->func.common.type & ZEND_USER_FUNCTION)
+			ir_ref type = ir_LOAD_U8(ir_ADD_OFFSET(func_ref, offsetof(zend_closure, func.type)));
+			if_cond_user = ir_IF(ir_AND_U8(type, ir_CONST_U8(ZEND_USER_FUNCTION)));
+			ir_IF_TRUE(if_cond_user);
+		}
 
-		// JIT: zend_jit_init_func_run_time_cache_helper(closure->func);
-		ir_CALL_1(IR_VOID, ir_CONST_FC_FUNC(zend_jit_init_func_run_time_cache_helper),
-			ir_ADD_OFFSET(func_ref, offsetof(zend_closure, func)));
+		if (!func || func->common.type == ZEND_USER_FUNCTION) {
+			// JIT: if (closure->func.op_array.run_time_cache__ptr)
+			if_cond = ir_IF(ir_LOAD_A(ir_ADD_OFFSET(func_ref, offsetof(zend_closure, func.op_array.run_time_cache__ptr))));
+			ir_IF_FALSE(if_cond);
```

Comment:
Why do we need this additional check?
We checked ``!RUN_TIME_CACHE(op_array)`` in ``zend_jit_init_func_run_time_cache_helper()``.

Reply:
I see. The check for `!RUN_TIME_CACHE(op_array)` already existed and I missed that `zend_jit_init_func_run_time_cache_helper()` already made this check. We can remove the `!RUN_TIME_CACHE` check from the generated code. I'll commit that change.

On master, I see that `zend_jit_init_func_run_time_cache_helper()` _also_ checks function type (not on PHP-8.4), but I think it's preferable to check it in the generated code so that we can sometimes skip generating a call to `zend_jit_init_func_run_time_cache_helper()` when `func` is known to be an internal function.

Reply:
The function type check on master was introduced in 097edc86c820178da7b26c50bb87693b45e3be8e.
The only other place that calls `zend_jit_init_func_run_time_cache_helper()` already knows that this is a user function.
Perhaps it's even beneficial to create 2 helper variants of `zend_jit_init_func_run_time_cache_helper()`: one with a function type check and one without? Then we could use the one with the check in `zend_jit_push_call_frame` if `!func`.

Reply:
Please check, how this affects the generated JIT code size. E.g. function/tracing JIT on Symfony demo.
I suppose, this shouldn't make a big difference.

Reply:
I tried running Symfony demo on function JIT, got an assertion failure instead:
```
ir_base[737] is in use list of ir_base[1]
php: php-8.4/ext/opcache/jit/ir/ir_check.c:394: ir_check: Assertion `ok' failed.
```

On tracing JIT the difference is very very small:
```
tracing with check in IR code: 628824
tracing with check in helper (2 helper variants): 628808
```
Only a difference of 16! So probably not worth creating helper variants.